### PR TITLE
milestone six

### PR DESCRIPTION
**Pull request by rindrajosia**
Update the friendship controller, to avoid user/friend duplication

### CODE REVIEW

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,
Nice introduction to the milestone.
You did a good job so far but it's not sufficient to move forward.
Here are some issues noticed when reviewing your project

1. I'm connected as **Alice Bar**. I don't think it's good to see that at `http://localhost:3000/users/1`
Screenshot at 2020-06-29 13:26:58
I'm talking about the `Add Friend` aside of **Alice Bar**. It means that you want that user to send a friend request to himself. 

Look what happen when I click on the button `Add Friend`

Screenshot at 2020-06-29 13:30:29

The result is not good. **An user should not sent a friendship to himself**

I got this error when I wanted to reject the Friendship

Screenshot at 2020-06-29 13:31:42


2- I created a second account, called **Bob Bar**. Here is what I saw when I connect
Screenshot at 2020-06-29 13:35:07
I couldn't be possible because **Bob Bar** and **Alice Bar** are not friends. As requested [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png), Posts timeline includes posts written by logged-in user and their friends.

Fix that please.

3- Now, I sent a friendship request to **Alice Bar**
Screenshot at 2020-06-29 13:38:30
I think **Alice Bar** should appear in the section called _Friend_requests sent by me (Waiting for confirmation)_.
Maybe it's the reason why the rejected request failed.

4- Look at this now
Screenshot at 2020-06-29 13:43:36
**Alice Bar** sent a friendship request to **Bob Bar**(the connected user).
**Alice Bar** appears in both sections.

At the same time, here is what into your `Friendship` table
```rb
2.7.0 :001 > Friendship.all
  Friendship Load (0.8ms)  SELECT  "friendships".* FROM "friendships" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Friendship id: 5, user_id: 1, friend_id: 2, status: false, created_at: "2020-06-29 12:42:33", updated_at: "2020-06-29 12:42:33">, #<Friendship id: 6, user_id: 2, friend_id: 1, status: false, created_at: "2020-06-29 12:42:33", updated_at: "2020-06-29 12:42:33">]> 
2.7.0 :002 >
```
Why did you decide you create two lines at the friendship request instead of adding the second line after the friendship be accepted?

5- I make **Bob Bar** accept the friendship request
Screenshot at 2020-06-29 13:47:30

The issue now is the `Friendship` table content
```rb 
2.7.0 :001 > Friendship.all
  Friendship Load (0.5ms)  SELECT  "friendships".* FROM "friendships" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Friendship id: 6, user_id: 2, friend_id: 1, status: false, created_at: "2020-06-29 12:42:33", updated_at: "2020-06-29 12:42:33">, #<Friendship id: 5, user_id: 1, friend_id: 2, status: true, created_at: "2020-06-29 12:42:33", updated_at: "2020-06-29 12:46:23">]> 
2.7.0 :002 >
```

6- **ADD INTEGRATION TESTS!!!** It's one of the most important requirements

You can notice that only one of the both line has is status set to `true`.

Kindly fix all the issues and submit a new code review request
Thanks

Mael FOSSO

**Comment by muhenge** [comment]
In this branch, we added shoulda methods to test our associations but we have a lint error because of its configuration
in the file _spec/rails_helper.rb_
Screenshot from 2020-06-30 00-09-36
this is the configuration
Screenshot from 2020-06-30 00-19-29
the lint error returning

File: app/controllers/friendships_controller.rb [hunk 1]
```diff
@@ -12,16 +13,26 @@ def create
   end
 
   def update
+    @friend = Friendship.find(params[:id])
+    @ft = Friendship.create
+    @ft.user_id = current_user.id
+    @ft.friend_id = @friend.user_id
+    @ft.status = true
+    @ft.save
```
Comment (uryelah) [inline app/controllers/friendships_controller.rb]: - [ ] Like the last reviewer said, you should only save the reversed friendship after confirming that you managed to update the status of the first request to true.
What happens if there's an error updating @friendship and you already saved @ft?

First update @friendship and if that works you can save @ft.

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -1,16 +1,17 @@
 <section class="user-section">
   <h2><%= "Name: #{@user.name}" %>
-    <% if @friend %>
+    <% if @friend && !(@user.id.equal?current_user.id) %>
```
Comment (uryelah) [inline app/views/users/show.html.erb]: - [ ] Good job showing the button only when the user is not the current user. But let's organize your code a bit.
View logic should never be in the views, it should be in helper methods. Make sure all those if/else statements are inside helper methods.

**Review by uryelah — CHANGES_REQUESTED** [review]
Hi **Herve Ngenzi & Josia Rakotonarivo**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job improving your project! There are still some issues to revise before continuing to the next milestone but hang in there and read the comments below with care. You can do it! 💪 

# Required Changes 🏗️ 🐛 💡

- About the linting error: You can disable that rubocop rule for that file to avoid the error or use a different word than 'config' in the block.

**Comment by uryelah** [comment]
**Kindly fix the pointed out issues before requesting the next code review**

------------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Changes Required

Hi team :wave: 

Nice implementing the changes from the last review. However, there is still one change to make. 
Stickler is pointing out a linting error. Kindly look into it.

Kindly fix the linting error, then submit another code review. If it persists, please reach out.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

**Review by rOluochKe — APPROVED** [review]
Hello Team!
### Well done and for making changes from the previous reviews :clap: Congratulations :boom: :100: Your project has been approved.
### Merge and submit the project. Good luck with the next project.
- Happy coding!

**Pull request merged**